Add support for extended and recursive globs

mckern opened the pull request:
Ran into #35, and recursion would be extremely useful for my org. So I've added it.
I figure that `extglob` and `globstar` are pretty safe shopts since Bash 4 came out in 2009, but I've left some "safety valves" to ensure that we don't try to set those options against Bash 3 if we run into it.

Oh, I also ran `shfmt` and `shellcheck` against my work, so if there's little bits of whitespace drift those were introduced by tooling and they're safe.

mckern committed:
Add support for globstar

Ensures that Bash version 4 is at least 4 before enabling it.

mckern committed:
Make extglob and globstar configurable

Also adds tests to validate that these options behave as expected.

mckern committed:
Cut bash_version_check() back a little bit

Matching shell options against versions is good but since only two
options are exposed through the plugin config, we only need to check
for extglob and globstar.

mckern commented:
Noticed that there's no CI wired up to this plugin (which is _really_ funny in context), but the tests passed for me when I ran them locally:

```
$ docker run -it --rm -v "$PWD:/plugin:ro" buildkite/plugin-tester
run.bats
 ✓ Shellcheck a single file
 ✓ Shellcheck multiple files
 ✓ Shellcheck multiple files with starglob
 ✓ Shellcheck multiple files with extglob
 ✓ Shellcheck a single file with single option
 ✓ Shellcheck a single file with multiple options
 ✓ Shellcheck multiple files with multiple options
 ✓ Shellcheck failure

8 tests, 0 failures
```

I'll try running these in a bit against my fork.

mckern committed:
Fix typo in README

pzeballos commented:
Hey @mckern! This is really good! 🎉 We only noticed two things:
- When you set either option to `false`, options are still enabled. You should also add tests for these cases.
- Just to make sure the options are only turned on for this particular command (and it doesn't bleed out of scope) you should revert the option back to what they were before the end of the hook.

Thanks!

mckern commented:
@pzeballos can-do. I'll push up some commits after lunch. Thanks for taking a look.

> When you set either option to false, options are still enabled. You should also add tests for these cases.

~Oof, because they're defined as booleans, they're being rendered out as 1 or 0 in the inherited environment variable aren't they? That's on me for assuming they'd just be unset -- the test was only for null/not-null strings.~

> Just to make sure the options are only turned on for this particular command (and it doesn't bleed out of scope) you should revert the option back to what they were before the end of the hook.

~I'd assumed that each stage of a pipeline run happens within its own sub-shell (with its own scoped environment), and there's only one `command` stage in a run; _can_ this bleed through the post-command stages?~

mckern committed:
Fix broken variable test, update test cases

Environment variables defined as boolean values in the spec will be
cast to "1" or "0", so we should evaluate for value instead of
just evaluating for presence.
This also adds test cases, exercising what happens when each option is
explicitly disable.

mckern committed:
Fix booleans *again*, for real this time

Previous fix was operating under a false assumption; after doing a
little debugging, the Buildkite Agent expresses these values as the
strings "true" or "false". Our test should take that into account.

This is implied by json-schema.org's guidlines for boolean values but
it is not made clear in the Buildkite plugin documentation.

mckern committed:
Booleans, Pt. 3: Bool Harder

It looks like the values of these options are passed through literally
as environment variables with no casting whatsoever done. So we need
to accommodate for "true" and "1" and case insensitivity when figuring
out if shopts should be set. There's some test cases to flex the
negative scenarios, as well as the positive.

mckern commented:
@pzeballos test cases for enabled and disabled are updated and logic now accommodates the _soft_ nature of booleans as expressed via Bash environment variables.

Regarding resetting `shopt` values after the command hook is run, I built a test branch with a post-command hook to evaluate `shopt` values and they appear to be contained to the singular command stage. Since there can only be one command hook (unlike other stages), that seems sufficient to prevent `shopt` inheritance between stages to me -- is this a scenario you've encountered before or is this more of a "let's try to cover all the bases" scenario?

pzeballos commented:
> is this a scenario you've encountered before or is this more of a "let's try to cover all the bases" scenario?

indeed, it was that. We'll review your update and let you know. Thanks for the quick turnaround @mckern!

pzeballos commented on hooks/command:
```diff
@@ -1,35 +1,80 @@
+    *)  fail "attempted to set unknown shopt '${option}'" ;;
+  esac
+
+  if [[ ${BASH_VERSINFO[0]} -lt ${required} ]]; then
+    fail "shopt '${option}' requires Bash ${required} or newer"
+  fi
```
Can you add tests for each option to verify the behavior when the BASH_VERSINFO is not enough?

mckern replied:
`BASH_VERSINFO` is an internal bash variable and I'm not sure it can be overriden. Might require experimentation and/or getting "creative".

mckern replied:
Yeah, internal bash variables are read-only so this will require shimming in an override if we want to test this:

```
$ bash -x ./test.sh
+ BASH_VERSINFO=(1 2 3 4)
./test.sh: line 3: BASH_VERSINFO: readonly variable
+ echo '5 1 16 1 release x86_64-apple-darwin20.6.0'
5 1 16 1 release x86_64-apple-darwin20.6.0
```

mckern replied:
OK -- so thinking about this for a few minutes, I think the paths for testing this boil down to:

- Stub in version assignment to allow overriding `BASH_VERSINFO` with a differently named debugging variable during this specific function during testing; effective, but kind of clunky and risky since that override variable can also be set during CI runs (which, like, why would anyone do that? I don't know but it'd be _possible_).
- Modify the `buildkite/plugin-tester` testing container that y'all maintain to allow for multiple versions for Bash to be run. A neat idea, but beginning to approach "ocean boiling" levels of effort.
- Evaluate it locally because it's a pretty straightforward bit of logic comparing an internal bash variable that will only ever be set directly by the actual shell that the plugin is running in (and never, ever inherited because it _can't_ be).

```
$ /bin/bash --version
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin21)
Copyright (C) 2007 Free Software Foundation, Inc.
$ /usr/local/bin/bash --version
GNU bash, version 5.1.16(1)-release (x86_64-apple-darwin20.6.0)
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
$ BUILDKITE_PLUGIN_SHELLCHECK_GLOBSTAR=true BUILDKITE_PLUGIN_SHELLCHECK_FILES="**/*.sh" /bin/bash hooks/command
ERROR: shopt 'globstar' requires Bash 4 or newer
$ BUILDKITE_PLUGIN_SHELLCHECK_GLOBSTAR=true BUILDKITE_PLUGIN_SHELLCHECK_FILES="**/*.sh" /usr/local/bin/bash hooks/command
+++ Running shellcheck on 4 files

In tests/testdata/subdir/llamas.sh line 2:
echo $UNDEFINED_VAR
     ^------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
echo "$UNDEFINED_VAR"

For more information:
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
$
```

How would you like to proceed?

mckern replied:
OK, there's a 4th option that occurs to me and is decoupled from the version check: don't check the version, just attempt to set the `shopt` and gate on the success/failure of that:

```
$ shopt -s globstar
-bash: shopt: globstar: invalid shell option name
$ shopt -s globstar &>/dev/null; echo $?
1
```

pzeballos replied:
This is fantastic @mckern! Great investigation, thank you very much! 🙏🏻 You are right, and this is practically impossible to test 😅. Your 4th option would be fantastic, but because `shopt` it's a builtin I don't think it would be possible.
I think at this time, we could just probably add a comment to that part of the code with the warning to be careful.

mckern replied:
Just pushed a refactored commit that no longer uses a strict bash version check and includes numerous notes about what's going on under the hood.

```
$ git describe --always --long
ab24db5
$ BUILDKITE_PLUGIN_SHELLCHECK_GLOBSTAR=true BUILDKITE_PLUGIN_SHELLCHECK_FILES="**/*.sh" /usr/local/bin/bash hooks/command
+++ Running shellcheck on 4 files

In tests/testdata/subdir/llamas.sh line 2:
echo $UNDEFINED_VAR
     ^------------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
echo "$UNDEFINED_VAR"

For more information:
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
$ BUILDKITE_PLUGIN_SHELLCHECK_GLOBSTAR=true BUILDKITE_PLUGIN_SHELLCHECK_FILES="**/*.sh" /bin/bash hooks/command
ERROR: failed to set shopt 'globstar: shopt 'globstar' requires Bash 4 or newer
$ docker run -it --rm -v "$PWD:/plugin:ro" buildkite/plugin-tester
run.bats
 ✓ Shellcheck a single file
 ✓ Shellcheck multiple files
 ✓ Shellcheck multiple files using recursive globbing enabled with 'true'
 ✓ Shellcheck multiple files using extended globbing enabled with '1'
 ✓ Recursive globbing fails if starglob is disabled with 'false'
 ✓ Extended globbing fails if extended globbing is disabled with 'FALSE'
 ✓ Extended globbing fails if extended globbing is disabled with '0'
 ✓ Shellcheck a single file with single option
 ✓ Shellcheck a single file with multiple options
 ✓ Shellcheck multiple files with multiple options
 ✓ Shellcheck failure

11 tests, 0 failures

$
```

pzeballos commented on tests/run.bats:
```diff
@@ -36,6 +36,68 @@ load '/usr/local/lib/bats/load.bash'
   unstub docker
 }
 
+@test "Shellcheck multiple files using recursive globbing enabled with '1'" {
```
```suggestion
@test "Shellcheck multiple files using recursive globbing enabled with 'true'" {
```

pzeballos commented on tests/run.bats:
```diff
@@ -36,6 +36,68 @@ load '/usr/local/lib/bats/load.bash'
+  unstub docker
+}
+
+@test "Shellcheck multiple files using extended globbing enabled with 'true'" {
```
```suggestion
@test "Shellcheck multiple files using extended globbing enabled with '1'" {
```

mckern committed:
Update tests/run.bats

Co-authored-by: Pol <pzeballos@users.noreply.github.com>

mckern committed:
Update tests/run.bats

Co-authored-by: Pol <pzeballos@users.noreply.github.com>

mckern committed:
Remove version note for extglob

`extglob` was introduced in Bash 2, which had its last release in
2002. It's a pretty safe target, so we don't need a note about it.

mckern committed:
Refactor testing for valid shopts

Instead of testing against a version lookup, we can just try setting
the shopt directly and handling the failure ourselves. It's a little
more brute-force but it's also a little bit less dependent on internal
Bash variables.

The nature of shopts means we can't directly test them without
overriding so much code that we're essentially using a function to
exercise a test (not the other way around). But there's ample comments
describing the important details, like "no major platform ships bash
3 except macOS".

mckern committed:
typo fixup

mckern commented on plugin.yml:
```diff
@@ -6,6 +6,12 @@ requirements:
   - docker
 configuration:
   properties:
+    globstar:
+      type: boolean
+      default: false
+    extglob:
+      type: boolean
+      default: false
```
Thinking about this a little bit: do we want to directly expose the implementation detail by using the `shopt` names, or would more generic keys like `extended_regex` and `recursive_regex` (or similar) be more appropriate?

toote replied:
that is actually an interesting idea that would probably simplify the code a lot turning on both options if available (and adding some kind of output with a notice about it)

mckern replied:
Pushed it up as a new commit, with a new test and the corresponding feedback about notifying users that behavior has been modified.

mckern committed:
Use nicer keys to enable glob modifiers

Instead of directly exposing implementation details by using Bash
shopts as config keys, use the nicer `extended_glob` and
`recursive_glob` keys.

Evaluating for the expected variables moves out into the main script
body, and casting the variable values to boolean values is now its own
function. Setting either value will also now emit an explicit message
to stdout, underneath their own (de-emphasized) build-phase grouping.

A test has been added to validate that expected behavior still applies
when both options are enabled at the same time.

toote reviewed (changes requested):
Looking great... but we have fixed the tests in the `master` branch, which means your tests are now failing (the same way they should have been failing before).

I tried making the change myself but I didn't have permissions to do so. After the mergeback from this is the diff:
```diff
diff --git a/tests/run.bats b/tests/run.bats
index cabec31..c1fdad6 100644
--- a/tests/run.bats
+++ b/tests/run.bats
@@ -43,7 +43,7 @@ load '/usr/local/lib/bats/load.bash'
   export BUILDKITE_PLUGIN_SHELLCHECK_FILES="**/*.sh"
 
   stub docker \
-    "run --rm -v $PWD:/mnt --workdir /mnt koalaman/shellcheck:stable --color=always \"tests/testdata/recursive/subdir/stub.sh tests/testdata/test.sh tests/testdata/subdir/llamas.sh tests/testdata/subdir/shell with space.sh\"' : echo testing stub.sh test.sh llamas.sh shell with space.sh"
+    "run --rm -v $PWD:/mnt --workdir /mnt koalaman/shellcheck:stable --color=always tests/testdata/recursive/subdir/stub.sh tests/testdata/subdir/llamas.sh tests/testdata/subdir/shell\ with\ a\ space.sh tests/testdata/test.sh : echo testing stub.sh test.sh llamas.sh shell with space.sh"
 
   run "$PWD/hooks/command"
 
@@ -59,7 +59,7 @@ load '/usr/local/lib/bats/load.bash'
   export BUILDKITE_PLUGIN_SHELLCHECK_FILES="tests/testdata/subdir/*.+(sh|bash)"
 
   stub docker \
-    "run --rm -v $PWD:/mnt --workdir /mnt koalaman/shellcheck:stable --color=always \"tests/testdata/subdir/llamas.sh tests/testdata/subdir/shell with space.sh tests/testdata/subdir/stub.bash\"' : echo testing llamas.sh shell with space.sh stub.bash"
+    "run --rm -v $PWD:/mnt --workdir /mnt koalaman/shellcheck:stable --color=always tests/testdata/subdir/llamas.sh tests/testdata/subdir/shell\ with\ a\ space.sh tests/testdata/subdir/stub.bash : echo testing llamas.sh shell with space.sh stub.bash"
 
   run "$PWD/hooks/command"
 
@@ -106,7 +106,7 @@ load '/usr/local/lib/bats/load.bash'
   export BUILDKITE_PLUGIN_SHELLCHECK_FILES="**/*.+(sh|bash)"
 
   stub docker \
-    "run --rm -v $PWD:/mnt --workdir /mnt koalaman/shellcheck:stable --color=always \"tests/testdata/recursive/subdir/stub.bash tests/testdata/subdir/stub.bash tests/testdata/recursive/subdir/stub.sh tests/testdata/test.sh tests/testdata/subdir/llamas.sh tests/testdata/subdir/shell with space.sh\"' : echo testing stub.sh test.sh llamas.sh shell with space.sh"
+    "run --rm -v $PWD:/mnt --workdir /mnt koalaman/shellcheck:stable --color=always tests/testdata/recursive/subdir/stub.bash tests/testdata/recursive/subdir/stub.sh tests/testdata/subdir/llamas.sh tests/testdata/subdir/shell\ with\ a\ space.sh tests/testdata/subdir/stub.bash tests/testdata/test.sh : echo testing stub.sh test.sh llamas.sh shell with space.sh"
 
   run "$PWD/hooks/command"
```

toote commented:
> Noticed that there's no CI wired up to this plugin (which is _really_ funny in context)

Actually, it is wired, but turned off for third-party forks

mckern committed:
Fix tests after PR #44

Co-authored-by: Matías A. Bellone <matias.bellone@builkite.com>

mckern commented:
@toote thanks for the heads up; I applied your diff and committed it with you as co-author.

toote reviewed (approved):
Worked like a charm :)

mckern commented on hooks/command:
```diff
@@ -1,35 +1,94 @@
+  local value="${1}"
+  shopt -s nocasematch
+
+  # return immediately (and successfully) unless the value is explicitly set to true
```
```suggestion
  # return immediately (and successfully) if the value is explicitly set to true
```

mckern committed:
Fix typo

pzeballos reviewed (approved):
Thank you for your contributions and bearing with us! 🙏🏻
